Add `-no-toolchain-stdlib-rpath` flag.

dan-zheng opened the pull request:
Add `-no-toolchain-stdlib-rpath` flag: the negative version of
`-toolchain-stdlib-rpath`.

Make `-no-toolchain-stdlib-rpath` be the default: use `/usr/lib/swift` as
default RPATH on Darwin platforms instead of toolchain standard library.

---

Adapted from https://github.com/apple/swift/pull/27206.

tensorflow branch requires the opposite default (use toolchain standard
library as RPATH) because some stdlib modules like TensorFlow do not exist in
`/usr/lib/swift`.

dan-zheng committed:
Add `-no-toolchain-stdlib-rpath` flag.

Add `-no-toolchain-stdlib-rpath` flag: the negative version of
`-toolchain-stdlib-rpath`.

Make `-no-toolchain-stdlib-rpath` be the default: use `/usr/lib/swift` as
default RPATH on Darwin platforms instead of toolchain standard library.

Adapted from https://github.com/apple/swift/pull/27206.

tensorflow branch requires the opposite default (use toolchain standard
library as RPATH) because some stdlib modules like TensorFlow do not exist in
`/usr/lib/swift`.

dan-zheng commented:
Patch suggested by @brentdax, to minimize the diff between `master` and `tensorflow` branches.
I'm not sure how to add tests, suggestions welcome.

brentdax commented:
To test this, I would add test cases to `test/Driver/linker-rpath.swift` which duplicate the existing `x86_64-apple-macosx10.14` and `x86_64-apple-macosx10.15` tests but pass `-no-toolchain-stdlib-rpath`. Your change in the TensorFlow branch should probably require you to modify the originals of these test cases; just leave the copies testing the original results.

dan-zheng committed:
Add `-no-toolchain-stdlib-rpath` tests to test/Driver/linker-rpath.swift.

dan-zheng commented:
> To test this, I would add test cases to `test/Driver/linker-rpath.swift` which duplicate the existing `x86_64-apple-macosx10.14` and `x86_64-apple-macosx10.15` test cases but pass `-no-toolchain-stdlib-rpath`. Your change in the TensorFlow branch should probably require you to modify the originals of these test cases; just leave the copies testing the original results.

Done in https://github.com/apple/swift/pull/27207/commits/9a1b3f2b7608d480f9470061cfa9359dbe23ea4d.

I added `-no-toolchain-stdlib-rpath` test cases for `x86_64-apple-macosx10.9` and `x86_64-apple-macosx10.15`, following existing tests in the same file for `-no-stdlib-rpath` and `-toolchain-stdlib-rpath`.

I added mirror tests in the `tensorflow` branch patch: https://github.com/apple/swift/pull/27206/commits/1ad2aa45013396810b1590d3a49e04bbfab62586.

jrose-apple commented on include/swift/Option/Options.td:
```diff
@@ -508,6 +508,9 @@ def no_static_stdlib: Flag<["-"], "no-static-stdlib">,
   HelpText<"Add an rpath entry for the toolchain's standard library, rather than the OS's">;
+def no_toolchain_stdlib_rpath: Flag<["-"], "no-toolchain-stdlib-rpath">,
+  Flags<[HelpHidden,DoesNotAffectIncrementalBuild]>,
+  HelpText<"Do not add an rpath entry for the toolchain's standard library, rather than the OS's">;
```
This help text feels weird because of that "rather" clause (copied from the original option). Does it mean the OS's rpath will be added? (No, only if necessary.)

I'm having trouble coming up with a better phrasing, but there probably is one.

brentdax replied:
"Do not add an rpath entry for the toolchain's standard library" alone is probably enough, perhaps with the word "(default)" as well.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/27207/commits/adce3fb793e80cbd6fb87252f3f7287f95d86d90:

```
"Do not add an rpath entry for the toolchain's standard library (default)"
```

jrose-apple reviewed (commented):
This seems all right to me. I'm not quite sure what the future of `-toolchain-stdlib-rpath` mode is on Apple platforms at all, but it seems like the same things we need for testing the master stdlib also apply to using the TensorFlow stdlib.

(A bunch of this might get reorganized too as part of the "generalize cross-compilation and SDK layouts" work described [on the forums](https://forums.swift.org/t/swift-linux-layout-considerations-aka-linux-is-difficult-lets-go-shopping/27400/18). But for now this is good.)

beccadax reviewed (approved):
We should settle the help text question, but otherwise this LGTM.

dan-zheng committed:
Improve help text for `-no-toolchain-stdlib-rpath` flag.

dan-zheng commented:
> This seems all right to me. I'm not quite sure what the future of `-toolchain-stdlib-rpath` mode is on Apple platforms at all, but it seems like the same things we need for testing the master stdlib also apply to using the TensorFlow stdlib.
> 
> (A bunch of this might get reorganized too as part of the "generalize cross-compilation and SDK layouts" work described [on the forums](https://forums.swift.org/t/swift-linux-layout-considerations-aka-linux-is-difficult-lets-go-shopping/27400/18). But for now this is good.)

Thank you for the helpful context.

---

Thanks to reviewers!

dan-zheng commented:
@swift-ci Please test
